### ISSUE TITLE
Run daemon in its own docker container

### ISSUE THREAD

**Issue by player-two** [newrelic/c-sdk#6]
Hey there!  I was a user of the [Agent SDK](https://docs.newrelic.com/docs/introduction-new-relic-agent-sdk) and noticed the deprecation notice, then found my way here.  One important difference in functionality is the (lack of) embedded mode.

For context, the application in question is a http server running in a docker container in k8s.  I typically adhere to the best practice of running a single process per container, so I would prefer not to run the daemon in the same container as the application.  That in turn means the app needs to talk to the daemon over a network rather than a unix domain socket.

Though I'm no k8s expert, I believe the networking model can be configured such that containers are private within a SDN, so in my specific case this should be ok security-wise, even without any authentication mechanism.  Though it may be best to attach a warning to this feature as it should be used with caution.

Is this something you're open to supporting?  I got the code working (see the linked PR), but I thought it would be best to have a conversation here before paying too much attention to it.

**Comment by crenshaw** [comment]
I'm glad you found your way here.  Thanks so much for the issue and the accompanying PR.

I'm an engineer on the team that develops the C SDK.  I appreciate that your issue is pointing out an important thing that the C SDK is missing.  For it to make sense in modern deployments it needs the ability to execute the daemon in its own container.  Instead of just communicating over loopback, the daemon needs to be enhanced so that it can talk over a socket. Making this kind of enhancement to the C SDK is definitely something we plan to do in the next few months.

I have two asks:

1. I see that your draft PR uses a bit of host:port parsing to get the job done.  I also see that you offer a bit of hostname resolution for some polish.  Are there other requirements that you might need to help you to deploy the daemon as a single process in a container?  I'd like very much to learn about your use case.

2. For this particular OSS project, we aren't developing *in the open* yet, so we won't be able to accept your PR as-is and merge it here.  In house, the C SDK is part of a larger repo that hasn't been made available as OSS yet.  So I'm hopeful you are patient with us in our discussions.  There will be a day when we close your PR without merging it.  But we'll take care to give you a rad shout-out in our release notes at docs.newrelic.com and in this repo's CHANGELOG.md

**Comment by player-two** [comment]
1. Yeah, the parsing is (imo) easily the grossest part of that code.  The logic in there also got a little complex because I was trying not to change the function signatures or add another init function, e.g. `newrelic_init_remote()`.  That said, it wouldn't bother me as a user if there _was_ a different init function, I just didn't want to grow your surface area unnecessarily.

2. Not a problem at all - I care more that it gets done than how.  The only disadvantage for me is not being able to track progress since the repo is internal for y'all.  So if instead you can post an occasional update here and set expectations on timing, that would be lovely!


Diving more into my expected usage - I want it to work locally with Docker Compose and in a k8s cluster for higher environments.  [Compose's networking](https://docs.docker.com/compose/networking/) lets you specify a container name as a host, which is why the hostname resolution is helpful.  For example:

```yaml
version: '3'
services:
  web-app:
    # image/build info
  newrelic-daemon:
    image: newrelic-daemon
    command: ["newrelic-daemon", "-f", "--logfile", "/dev/stdout", "--port", "8080"]
```

My web app can then connect to `newrelic-daemon:8080`.  It doesn't matter much to me if I pass that to the sdk function as a single string, separate arguments, etc.

k8s provides a similar facility via [Service Networking](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/), so the hostname would look like `newrelic-daemon.<namespace>.svc.cluster.local`.

In the higher environments, I will be running a few replicas of the app but only a single instance of the daemon.  I haven't seen anything so far to indicate that will be a problem - I believe the golang http server in the standard library creates a new goroutine per request.

Hope the wall of text is helpful :)

**Comment by crenshaw** [comment]
Hello!    

I wanted to let you know that we are actively working on this issue.  First we'll get this working for the PHP Agent and then extend the new feature for the C-SDK.  They share a lot of internals.

My team is really grateful to you for your work on this issue.  I am particularly grateful for your feedback around how you'd like to be able to address the daemon, specifying its name in your Docker compose file and then connecting to `newrelic-daemon:8080`.  

We look forward to releasing it before the end of the year.

**Comment by Fahmy-Mohammed** [comment]
Implemented in [C-SDK version 1.2.0](https://github.com/newrelic/c-sdk/commit/e725025214bc0f5b2dc1e8e4d852f9b107d13e99) release.
